trillian/migrillian/core/controller.go

```diff
@@ -228,30 +228,72 @@ func (c *Controller) runWithRestarts(ctx context.Context) error {
 // have been transferred (in non-Continuous mode).
 func (c *Controller) Run(ctx context.Context) error {
 	metrics.controllerStarts.Inc(c.label)
+	dur := randDuration(c.opts.StopAfter, c.opts.StopAfter)
+	start := time.Now()
 
-	root, err := c.plClient.getVerifiedRoot(ctx)
+	// Note: Non-continuous runs are not affected by StopAfter.
+	pos, err := c.fetchTail(ctx, 0)
 	if err != nil {
 		return err
 	}
-	if c.opts.Continuous { // Ignore range parameters in Continuous mode.
-		c.opts.StartIndex, c.opts.EndIndex = int64(root.TreeSize), 0
-		glog.Warningf("%s: updated entry range to [%d, INF)", c.label, c.opts.StartIndex)
-	} else if c.opts.StartIndex < 0 {
-		c.opts.StartIndex = int64(root.TreeSize)
-		glog.Warningf("%s: updated start index to %d", c.label, c.opts.StartIndex)
+	if !c.opts.Continuous {
+		return nil
+	}
+
+	for dur == 0 || time.Since(start) < dur {
+		// TODO(pavelkalinnikov): Integrate runWithRestarts here.
+		next, err := c.fetchTail(ctx, pos)
+		if err != nil {
+			return err
+		}
+		if next == pos {
+			// TODO(pavelkalinnikov): Pause with accordance to the rate of growth.
+			// TODO(pavelkalinnikov): Make the duration configurable.
+			if err := clock.SleepContext(ctx, 30*time.Second); err != nil {
+				return err
+			}
+		}
+		pos = next
+	}
+
+	return nil
+}
+
+// fetchTail transfers entries within the range specified in FetcherConfig,
+// with respect to the passed in minimal position to start from, and the
+// current tree size obtained from an STH.
+func (c *Controller) fetchTail(ctx context.Context, begin uint64) (uint64, error) {
+	root, err := c.plClient.getVerifiedRoot(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	fo := c.opts.FetcherOptions
+	if fo.Continuous { // Ignore range parameters in continuous mode.
+		fo.StartIndex, fo.EndIndex = int64(root.TreeSize), 0
+		// Use non-continuous Fetcher, as we implement continuity in Controller.
+		fo.Continuous = false
```

Comment:
Is that going to be removed? If so can we do it now?

Reply:
Doing so will change how `Fetcher` behaves, so let's leave it for now. It is possible to remove, but for that I need to add another `Continuous` field specific to `Controller` config, in addition to `Fetcher`'s one.

Reply:
Can you make it a TODO then.

Reply:
Done.